src/arcsf/eval/utils.py

```diff
@@ -0,0 +1,299 @@
+                eval_rouge_recall(gen_output=generated_text, ground_truth=target_text)[
+                    "rougeL_recall"
+                ]
+            )
+
+        # get ground truth loss
+        gt_loss = get_loss(gt_outputs.logits, gt_batch["labels"])
+        output_dict["all_losses"][batch_start_index:batch_end_index, 0] = gt_loss.cpu()
+        # loop over perturbed samples to get their losses
+        for perturbed_index in range(1, n_perturbed + 1):
+            p_batch = logit_batch[perturbed_index]
+            with torch.no_grad():
+                p_output = model(**p_batch)
+            output_dict["all_losses"][
+                batch_start_index:batch_end_index, perturbed_index
+            ] = get_loss(p_output.logits, p_batch["labels"]).cpu()
+
+    # calculate truth_ratio and return them along with losses
+    output_dict["truth_ratios"] = truth_ratio(output_dict["all_losses"])
+
+    return output_dict
+
+
+def qualitative_eval(
+    model: torch.nn.Module,
+    tokenizer: AutoTokenizer,
+    dataset: EvalQADataset,
+    n_inputs: int,
+    random_seed: int,
+    **generate_kwargs: dict,
+) -> None:
+    """Performs qualitative evaluation of the selected model over selected data.
+    Prints the generated text given a question followed by the ground truth target.
+
+    Args:
+        model : Transformers model used to perform evaluation on
+        tokenizer : Tokenizer for the purpose of decoding model output
+        dataset : Dataset to perform evaluation on
+        n_inputs : number of inputs to sample
+        random_seed : Random seed for the dataloader
+        generate_kwargs : keyword arguments for the model.generate() method
+    """
+    # create dataloader with dataset
+    gen = torch.Generator().manual_seed(random_seed)
+    data_loader = DataLoader(dataset, batch_size=1, shuffle=True, generator=gen)
+
+    # loop over batches
+    for batch_idx, (question, answer) in enumerate(data_loader):
+        # get input and target
+        input_question = tokenizer.decode(question["input_ids"][0][0])
+        target = tokenizer.decode(answer["input_ids"][0][0])
+        # get model output
+        output = model.generate(
+            question["input_ids"][0],
+            attention_mask=question["attention_mask"][0],
+            pad_token_id=tokenizer.eos_token_id,
+            **generate_kwargs,
+        )
+        generated_text = tokenizer.decode(
+            output[0][len(question["input_ids"][0][0]) :], skip_special_tokens=True
+        )
+        # compare all three
+        print(f"\nQuestion index: {batch_idx}")
+        print(f"Question: {input_question}")
+        print(f"Generated: {generated_text}")
+        print(f"Target: {target}")
+
+        if batch_idx >= n_inputs:
+            break
+
+
+def get_analysis_values(
+    model_dir: str,
+) -> dict[np.ndarray, np.ndarray, np.ndarray, torch.Tensor, torch.Tensor]:
+    """
```

Comment:
```suggestion
) -> dict[str, np.ndarray | torch.Tensor]:
```

Reply:
I actually thought I set all of these to torch.Tensors. I'll have a look at that, but will add this change